tests/net/adversary.rs

```diff
@@ -7,14 +7,291 @@
+    ///
+    /// Instead of a handle, returns the node directly, as the adversary gets full access to all
+    /// nodes in the network.
+    #[inline]
+    pub fn faulty_nodes_mut(&mut self) -> impl Iterator<Item = &mut Node<D>> {
+        self.0.faulty_nodes_mut()
+    }
+
+    /// Returns a mutable node handle iterator over all nodes in the network.
+    #[inline]
+    pub fn correct_nodes_mut(&mut self) -> impl Iterator<Item = NodeMutHandle<D>> {
+        self.0.correct_nodes_mut().map(NodeMutHandle::new)
+    }
+
+    /// Normally dispatch a message
+    pub fn dispatch_message(&mut self, msg: NetMessage<D>) -> Result<Step<D>, CrankError<D>> {
+        self.0.dispatch_message(msg)
+    }
+
+    /// Injects a message into the network.
+    ///
+    /// Allows the injection of `msg` at `position` into the message queue.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `msg.from` is not a faulty or either `msg.from` or `msg.to` do not exist. Panics
+    /// if `position` is equal to `Before(idx)`, with `idx` being out of bounds.
+    #[inline]
+    pub fn inject_message(&mut self, position: QueuePosition, msg: NetMessage<D>) {
+        // Ensure the node is not faulty.
+        assert!(
+            self.0
+                .get(msg.from.clone())
+                .expect("inject: unknown sender node")
+                .is_faulty(),
+            "Tried to inject message not originating from a faulty node."
+        );
+
+        // Sender must exist.
+        self.0
+            .get(msg.to.clone())
+            .expect("inject: unknown recipient node");
+
+        // Insert into queue. `insert` will panic on out-of-bounds.
+        match position {
+            QueuePosition::Front => self.0.messages.push_front(msg),
+            QueuePosition::Back => self.0.messages.push_back(msg),
+            QueuePosition::Before(idx) => self.0.messages.insert(idx, msg),
+        }
+    }
+
+    /// Swap two messages in the message queue.
+    ///
+    /// # Panics
+    ///
+    /// Panics if either `i` or `j` are out-of-bounds.
+    #[inline]
+    pub fn swap_messages(&mut self, i: usize, j: usize) {
+        self.0.swap_messages(i, j);
+    }
+
+    /// Reorder all messages.
+    ///
+    /// Sorts all message with a comparator function.
+    ///
+    /// Sorting is not cheap, but not prohitibively so, since message queues tend to be small for
+    /// most test cases. See `VirtualNet::sort_messages_by` for notes about sorting efficiency.
+    #[inline]
+    pub fn sort_messages_by<F>(&mut self, f: F)
+    where
+        F: FnMut(&NetMessage<D>, &NetMessage<D>) -> cmp::Ordering,
+    {
+        self.0.sort_messages_by(f)
+    }
+}
+
+// Downgrade-conversion.
+impl<'a, D: DistAlgorithm> From<NetMutHandle<'a, D>> for NetHandle<'a, D> {
+    #[inline]
+    fn from(n: NetMutHandle<D>) -> NetHandle<D> {
+        NetHandle(n.0)
+    }
+}
+
+/// Immutable node handle.
+pub struct NodeHandle<'a, D: 'a>(&'a Node<D>)
```

Comment:
Is there a reason why you used `where` here instead of:

```rust
pub struct NodeHandle<'a, D: 'a, D: DistAlgorithm>(&'a Node<D>);
```

Reply:
Same question goes for all the other places where the `where` keyword is separated out from the `<'a, D: 'a>` trait and lifetime bounds.

Reply:
Same question applies for everywhere the `where D: DistAlgorithm` is separated from the `<'a, D: 'a>` bounds.

Reply:
The main issue is that things tend to grow and most functions/structs have additional requirements (typically things like `D::NodeId: Clone`.

The Lifetimes make it extra weird. I think I'll probably do every*where*.